[docs] Add an integration guideline page for In-App purchases

amandeepmittal opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Closes ENG-10452

Also, follows  #26422

# How

<!--
How did you build this feature or fix this bug and why?
-->

Provide a list of libraries and/or tutorials for integrating an In-app purchase library.

Also, add redirects to handle InAppPurchase API reference for the latest version (since we are near the release of SDK 50 to become the latest, it's a good practice to have redirects beforehand).

Also, add a redirect message when redirecting from SDK 50: https://github.com/expo/expo/pull/26425#issuecomment-1905784041

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

Run docs locally and visit: http://localhost:3002/guides/in-app-purchases/ or see the [preview link](/guides/in-app-purchases/).

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

amandeepmittal committed:
[docs] Add an integration guideline page for In-App purchases

amandeepmittal commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,51 @@
+<BoxLink
+  title={
+    <>
+      <CODE>react-native-iap</CODE>
```
@keith-kurak, is it worth mentioning this library? In their [docs] or repo I do not see anything mentioned about Expo.

Also, are there any other libraries or resources we should mention here?

amandeepmittal commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,51 @@
+
+In-app purchases (IAP) are transactions within a mobile or desktop application where users can buy digital goods or additional features. This guide provides a list of popular libraries and tutorials for implementing IAP in your Expo app.
+
+> In-app purchase libraries require configuring custom native code. Native code is not configurable when using Expo Go. Instead, create a [development build](/develop/development-builds/introduction/), which will allow you to use a native library in your project.
```
```suggestion
> In-app purchase libraries require configuring custom native code. Native code is not configurable when using Expo Go. Instead, create a [development build](/develop/development-builds/introduction/), which allows using a native library in your project.
```

amandeepmittal committed:
Apply suggestions from code review

keith-kurak commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,51 @@
+      <CODE>react-native-purchases</CODE>
+    </>
+  }
+  description="An open-source framework that provides a wrapper around Google Play Billing, StoreKit and RevenueCat. This library is a client for RevenueCat's subscription and purchase tracking system."
```
```suggestion
  description="An open-source framework that provides a wrapper around Google Play Billing and StoreKit API's, as well as integration with RevenueCat services supporting in-app purchases. `react-native-purchases` enables product management, analytics, and simplified workflows for in-app purchase requirements that may extend beyond your client code, such as validating purchases on your application backend."
```
maybe not the best wording, but wanted to emphasize the receipt validation angle, that's got a lot of appeal even if you don't like the other stuff.

keith-kurak commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,51 @@
+      <CODE>react-native-iap</CODE>
+    </>
+  }
+  description="A React Native library to access In-app purchases capabilities in your Android and iOS apps."
```
```suggestion
  description="A React Native library providing an interface to the client-side native in-app purchase API's built into Android and iOS."
```
Trying to emphasize more that it's "just the wrapper".

keith-kurak commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,51 @@
+/>
+
+## Libraries
+
```
```suggestion

These libraries provide robust support for in-app purchase functionality, as well as out-of-the-box compatibility with Expo apps using [CNG](workflow/continuous-native-generation/). 
```
Some kind of intro as to why we're recommending these (they have config plugins + Expo-specific docs)

keith-kurak reviewed (commented):
feedback added!

amandeepmittal committed:
Apply suggestions from code review

Co-authored-by: Keith Kurak <keith.kurak@gmail.com>

amandeepmittal commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,52 @@
+
+## Libraries
+
+These libraries provide robust support for in-app purchase functionality, as well as out-of-the-box compatibility with Expo apps using [CNG](workflow/continuous-native-generation/). 
```
```suggestion
These libraries provide robust support for in-app purchase functionality and out-of-the-box compatibility with Expo apps using [CNG](workflow/continuous-native-generation/).

```

amandeepmittal commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,52 @@
+      <CODE>react-native-purchases</CODE>
+    </>
+  }
+  description="An open-source framework that provides a wrapper around Google Play Billing and StoreKit API's, as well as integration with RevenueCat services supporting in-app purchases. `react-native-purchases` enables product management, analytics, and simplified workflows for in-app purchase requirements that may extend beyond your client code, such as validating purchases on your application ...
+  href="https://github.com/RevenueCat/react-native-purchases"
+  Icon={GithubIcon}
+/>
```
```suggestion
  description={<><CODE>react-native-purchases</CODE> is an open-source framework that provides a wrapper around Google Play Billing and StoreKit APIs, as well as integration with RevenueCat services supporting in-app purchases. It enables product management, analytics, and simplified workflows for in-app purchase requirements that may extend beyond your client code, such as validating purchases on your app's backend.</>}
  href="https://github.com/RevenueCat/react-native-purchases"
  Icon={GithubIcon}
/>
```

amandeepmittal committed:
Apply suggestions from code review

amandeepmittal committed:
Merge branch 'main' into aman/eng-10452-after-removing-in-app-purchases-create-a-list-guide-where-we

amandeepmittal commented on docs/pages/guides/in-app-purchases.mdx:
```diff
@@ -0,0 +1,53 @@
+
+## Libraries
+
+These libraries provide robust support for in-app purchase functionality and out-of-the-box compatibility with Expo apps using [CNG](workflow/continuous-native-generation/).
```
```suggestion
The following libraries provide robust support for in-app purchase functionality and out-of-the-box compatibility with Expo apps using [CNG](workflow/continuous-native-generation/) and provide [Config Plugins](/config-plugins/introduction/) for seamless integration in your app.
```

amandeepmittal committed:
Update docs/pages/guides/in-app-purchases.mdx

amandeepmittal committed:
Update to RedirectNotification - a general purpose redirect notify component

amandeepmittal committed:
Update all references of RedirectNotification and add it to Use In-app purchase guide

amandeepmittal commented:
@Simek, based on your feedback, I've made the following additions
- Renamed `VersionedRedirectNotification` to `RedirectNotification`
- Updated `RedirectNotification` references to include the redirect message in the `mdx` file
![CleanShot 2024-01-23 at 16 16 32](https://github.com/expo/expo/assets/10234615/de426498-74d9-4df7-adee-7cb19380592a)

- Added `RedirectNotification` to In App Purchase guide for SDK 50 and above
![CleanShot 2024-01-23 at 16 14 12](https://github.com/expo/expo/assets/10234615/c34793b7-c34d-4d46-ba23-61fbbb0a7ba2)

Simek reviewed (approved):
This looks great, thanks! 👍 

P.S. Let's also tinker next sync if there are any other situations in which we could benefit from using this redirect notice.
